### PR TITLE
Move storage integrations to be configured as services (PP-95)

### PR DESCRIPTION
**Pull request by jonathangreen**
## Description

- Add a dependency injection container, that contains services used by our controllers.
- Create framework for adding new services, and configuring them via environment variables, validated via pydantic.
- Remove the existing mirror integration code, and replace it with a new storage service.
  - Modify the marc exporter and s3 analytics provider to use the new storage service.
  - Add the new environment variables for the storage service to README.

Notes 🚨: (these are now all resolved)
- ~~The PR this is based on https://github.com/ThePalaceProject/circulation/pull/1358 should get merged before this one does.~~
- ~~This PR is dependant on an admin UI PR, that removes the admin UI side functionality for configuring storage integrations:  https://github.com/ThePalaceProject/circulation-admin/pull/87~~
- ~~Before this is merged we will need to update our hosting-playbook to deploy the relevant environment variables our to our hosting environments. PR for this up here: https://github.com/ThePalaceProject/hosting-playbook/pull/489~~

## Motivation and Context

- JIRA: PP-95
- The motivation here is to allow services that don't need to be configured in the admin UI to be configured via environment variables, and instantiated via a dependency injection container, so we are able to easily mock the services for testing.
- This uses the storage service as a test case for the new services architecture, and implements it as a s3 storage service.

## How Has This Been Tested?

Notes: The `docker-compose.yml` included in the repo was updated with the new environment variables, so that environment can be used for integration testing.

- Manually tested in a `docker-compose` environment.
  - Imported the Palace Bookshelf collection.
  - Configured s3 analytics and marc integrations.
  - Checked out a book.
    - Made sure analytics json files ended up in minio bucket.
  - Ran the cache_marc_files script.
    - Made sure that marc files ended up in the public access bucket.
    - Made sure links to the files worked at the `/marc` endpoint.
- Ran unit tests locally and in CI.

## Checklist

- [X] I have updated the documentation accordingly.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by jonathangreen** [comment]
This should be ready to merge now, once it gets a code review. All the dependant PRs have gone in, and its been rebased against the current `main` branch.

**Comment by tdilauro** [comment]
I started looking at this yesterday and will continue today.

File: scripts.py
Comment (tdilauro) [inline scripts.py:197]: I'm uncomfortable about having MagicMock in the app code. Why do we need it here? Are there reasonable alternatives?

**Review by tdilauro — COMMENTED** [review]
I'm still looking at the tests, but the app generally looks good. I do have a concern about relying on the `Dependency Injector` project because it's not clear that it is being maintained (they haven't merged a PR in 9 months).

File: api/s3_analytics_provider.py
Comment (jonathangreen) [reply api/s3_analytics_provider.py:229]: This is existing code, that was just moved around by this PR:
https://github.com/ThePalaceProject/circulation/blob/49a5d4bbb8cfca9e9e8b20f1f113b2382d0bf9c9/core/s3.py#L534-L574

Since it did not seem that determining these URLs belonged as part of the s3 class. I'd like to keep the original style intact here, so its clear the code was moved around when looking at diff / blame in the future.

File: core/marc.py
Comment (jonathangreen) [reply core/marc.py:645]: Same comment as above. This code duplication existed before this PR, the functions just moved between classes. I'd like to keep the original code intact until this PR goes in. We can factor this out to a helper if we want to after this lands.

File: core/marc.py
Comment (jonathangreen) [reply core/marc.py:675]: Resolved

File: core/mock_analytics_provider.py
Comment (jonathangreen) [reply core/mock_analytics_provider.py:4]: Resolved

File: core/model/collection.py
Comment (jonathangreen) [reply core/model/collection.py:951]: This code here is about to be deleted. 

The only reason I didn't remove it as part of this PR is make it easier to roll this work back if there is a problem. The follow up PR is open here https://github.com/ThePalaceProject/circulation/pull/1379. I don't think factoring this into a constant for the time between when this lands and when #1379 lands is worthwhile.

File: core/scripts.py
Comment (jonathangreen) [reply core/scripts.py:129]: Not sure I agree here, but I updated this one anyway.

File: core/service/storage/configuration.py
Comment (jonathangreen) [reply core/service/storage/configuration.py:10]: Okay. I removed the default here. This required a good number of code changes. I pushed a new commit that:
- Updates documentation
- Updates tests
- Adds new tests to s3 class constructor to make sure the URL template is consistent with the now optional region parameter.

File: scripts.py
Comment (jonathangreen) [reply scripts.py:197]: Good catch. I'm not sure what I was thinking here. I must have thought this was a test mock class. I updated it to use the scripts container property.

**Comment by jonathangreen** [comment]
I pushed a few commits resolving code review feedback.

> I do have a concern about relying on the Dependency Injector project because it's not clear that it is being maintained (they haven't merged a PR in 9 months).

Its hard to find a dependency injection container for Python. In the past I've found that constructor injection with a dependency injection container to wire up the dependencies is a helpful structure for webapps, and my hope is that we can slowly decouple our services this way.

Despite not having any PRs merged in the last 9 months, the library does have 1 million downloads a month on Pypy making it by far the most popular dependency injection container I've found for Python. It also has well written documentation https://python-dependency-injector.ets-labs.org/ and has been around since 2015. 

There is a issue here https://github.com/ets-labs/python-dependency-injector/issues/688 where the maintainer mentions that he considers the library feature complete at this point, but intends to support it to update python versions, dependency versions etc.

If you have a suggestion of a different library, I'm willing to take a look at switching this code. Otherwise the library becoming unsupported is a risk, but IMO the immediate gain of decoupling our services is worth it.

File: core/scripts.py
Comment (tdilauro) [reply core/scripts.py]: That's why it's "minor" and "style". I def won't be offended if you disagree or otherwise decide not to change it. I'll try to use "style pref" in the future, to make that more clear.

I generally like the ternary in these cases because it does two things:
1. Makes it more clear to me that the main thing happening is that `self._services` is gonna get something assigned, either a specified value or a default.
2. Makes it impossible for what's happening in ll. 127-128 and l. 129 above to get separated in the code.

(2) here is also why I'm a fan of the walrus operator, though I never suggest that anyone change that one, since it seems more controversial.

**Comment by jonathangreen** [comment]
I rebased this with main to resolve the conflicts with https://github.com/ThePalaceProject/circulation/pull/1268. The only changes in the rebase other then imports and poetry are in `tests/core/test_marc.py` around creating the fake search index for the marc tests, since that changed in #1268.

**Review by tdilauro — APPROVED** [review]
This looks good! 🚀 

One question about whether a test might need a network connection.

File: tests/core/service/storage/test_configuration.py
Comment (jonathangreen) [reply tests/core/service/storage/test_configuration.py:9]: I don't think so, but I'll verify.

Comment (jonathangreen) [reply tests/core/service/storage/test_configuration.py:9]: Doesn't look like it. It seems like the available regions get loaded from the huge amount of json data that is included with botocore.

**Pull request merged**
